Provide more useful messages for job failures

chapmanb opened the pull request:
Failing pipelines with an error within a job were quite difficult to
debug as they require digging into the jobStore to identify error
messages. This path improves the output to report the underlying failure
at the top level to more easily identify the problem.

Original failures look like:
```
Exception in thread Thread-1:
Traceback (most recent call last):
[...]
    raise RuntimeError("The termination flag is set, exiting")
RuntimeError: The termination flag is set, exiting

Traceback (most recent call last):
[...]
  File "/home/chapmanb/bio/toil/venv/local/lib/python2.7/site-packages/toil-3.3.0a1-py2.7.egg/toil/leader.py", line 69
7, in mainLoop
    raise FailedJobsException(config.jobStore, len(toilState.totalFailedJobs), toilState.totalFailedJobs)
toil.leader.FailedJobsException: The job store '/tmp/user/1001/tmplfzfpv' contains 1 failed jobs
```
which do not indicate the actual error and require looking in the job
store directory (at least for file system job stores), looking through
all jobs in the store, then looking through all files in the storage to
find the error.

After the fix we now extract the underlying error, reporting it along
with the failed job ID:
```
toil.leader.FailedJobsException: The job store '/tmp/user/1001/tmp5o3I44' contains 5 failed jobs: k/o/jobiDqKj4, 4/z/jobOYTUGp, Y/G/job2MEtR3, J/t/jobGM576Q, 0/I/jobgSvX9x
=========> Failed job k/o/jobiDqKj4
---TOIL WORKER OUTPUT LOG---
Traceback (most recent call last):
  File "/home/chapmanb/bio/toil/venv/local/lib/python2.7/site-packages/toil-3.3.0a1-py2.7.egg/toil/worker.py", line 332, in main
    job._runner(jobWrapper=jobWrapper, jobStore=jobStore, fileStore=fileStore)
  File "/home/chapmanb/bio/toil/venv/local/lib/python2.7/site-packages/toil-3.3.0a1-py2.7.egg/toil/job.py", line 2323, in _runner
    returnValues = self._run(jobWrapper, fileStore)
  File "/home/chapmanb/bio/toil/venv/local/lib/python2.7/site-packages/toil-3.3.0a1-py2.7.egg/toil/job.py", line 2271, in _run
    return self.run(fileStore)
  File "/home/chapmanb/bio/toil/venv/local/lib/python2.7/site-packages/toil-3.3.0a1-py2.7.egg/toil/cwl/cwltoil.py", line 235, in run
    **self.executor_options)
  File "/home/chapmanb/bio/toil/venv/local/lib/python2.7/site-packages/cwltool/main.py", line 203, in single_job_executor
    for r in jobiter:
  File "/home/chapmanb/bio/toil/venv/local/lib/python2.7/site-packages/cwltool/draft2tool.py", line 218, in job
    builder = self._init_job(joborder, **kwargs)
  File "/home/chapmanb/bio/toil/venv/local/lib/python2.7/site-packages/cwltool/process.py", line 407, in _init_job
    raise WorkflowException("Error validating input record, " + str(e))
WorkflowException: Error validating input record, could not validate field `echo_in` because
  the value `{'instr': 'one'}` is not string
could not validate field `first` because
  the value `[{'instr': 'one'}, {'instr': 'two'}, {'instr': 'three'}, {'instr': 'four'}]` is not string
Exiting the worker because of a failed jobWrapper on host kunkel
ERROR:toil.worker:Exiting the worker because of a failed jobWrapper on host kunkel
<=========
```

cgcloud-jenkins commented:
Can one of the admins verify this patch?

hannes-ucsc commented on src/toil/leader.py:
```diff
@@ -464,8 +464,18 @@ def getJob(jobId):
 class FailedJobsException( Exception ):
-    def __init__(self, jobStoreLocator, numberOfFailedJobs):
-        super( FailedJobsException, self ).__init__( "The job store '%s' contains %i failed jobs" % (jobStoreLocator, numberOfFailedJobs))
+    def __init__(self, jobStoreLocator, numberOfFailedJobs, failedJobs, jobStore):
```
This signature makes no sense. If you are passing failedJobs, you don't need to pass its length in a separate argument.

chapmanb commented:
Hannes -- thanks for the suggestion. I removed the redundant information passed to the error message. Let me know if you spot other problems.

hannes-ucsc commented:
Please squash.

Jenkins, test this please.

chapmanb commented:
Squashed. Y'all are really stretching the limits of my git skillz

fnothaft commented:
> Squashed. Y'all are really stretching the limits of my git skillz

It is a good day if you learn something new, tho.
;)

hannes-ucsc commented:
One more thing: There is actually a lot going on in the exception constructor. One thing to keep in mind is that when the exception is being created, things in Toil might be generally out of whack. By performing more complex operations in that situation, operations that could fail on their own, you risk masking the original problem/exception with a secondary one, which will make diagnosis harder. I propose that you wrap the code you added to the constructor in

```py
try:
…
except:
    log.exception('Exception when compiling information about failed jobs')
```

and make sure that the exception is created with a reasonable fallback-message even after the code failed to compile the job information.

hannes-ucsc commented:
Jenkins, test this please.

chapmanb commented:
Hannes -- thanks for the suggestion. I added that safety out and squashed again so it's a single commit.

Frank -- I'm honestly afraid of all this `git push -f` ing but if you prefer to pretend I never mess up and do everything perfect the first time I'll do my best to make it look like that.

fnothaft commented:
@chapmanb 
> Frank -- I'm honestly afraid of all this git push -f ing but if you prefer to pretend I never mess up and do everything perfect the first time I'll do my best to make it look like that.

Oh, I don't blame you. Whenever I do something in git that I don't entirely trust, I copy my work over to another branch first. You can work a lot of recovery magic with git's reflog, but nothing's quite as reassuring of a safety net as having a backup.

I'm wouldn't expect perfection though. Git's great, and it's really powerful, but I think the learning curve is uniquely steep, with such spectacular drop-offs...

hannes-ucsc commented:
Jenkins, test this please.
